src/lib.rs

```diff
@@ -28,28 +32,35 @@ use zenoh::{
     subscriber::Subscriber,
 };
 
-pub type UtransportListener = Box<dyn Fn(Result<UMessage, UStatus>) + Send + Sync + 'static>;
-
 const UATTRIBUTE_VERSION: u8 = 1;
 
 pub struct ZenohListener {}
 pub struct UPClientZenoh {
     session: Arc<Session>,
     // Able to unregister Subscriber
-    subscriber_map: Arc<Mutex<HashMap<String, Subscriber<'static, ()>>>>,
+    #[allow(clippy::type_complexity)]
+    subscriber_map:
+        Arc<Mutex<HashMap<UUri, HashMap<Arc<ListenerWrapper>, Subscriber<'static, ()>>>>>,
```

Comment:
I agree that _if we know what we're doing_ then what you propose is still fine to do.

In Zenoh once a callback is registered, we don't need to hang onto the `ListenerWrapper` anymore. I can understand that.

However, I do not wish to expose that hash from within `ListenerWrapper` because I'd like to keep the responsibilities clear to other implementors of `up-client-foo-rust` libraries.

In other protocols, they may need to keep track of the `ListenerWrapper` in order to call `on_receive()` on it.

So the intention here is to emphasize that if we wish to keep track of an instance of a concrete implementation of a UListener, `ListenerWrapper` is _the_ standard way to do that.

Note that with the way `up-rust` is implemented with `UListener` now, it's in-fact possible for a `up-client-foo-rust` to roll their own `MyCoolListenerWrapper` which does expose that hash if needed, but I'd like to avoid that being the "commonly accepted" path.

Reply:
My main idea is to avoid using two HashMap here. We need to use `UUri` to find the `ListenerWrapper` and then use that `ListenerWrapper` to find the Zenoh callback.
Therefore, I wish we can use one HashMap. Use both `UUri` + `ListenerWrapper` as key, and find the value (Zenoh callback).

> However, I do not wish to expose that hash from within ListenerWrapper because I'd like to keep the responsibilities clear to other implementors of up-client-foo-rust libraries.

I agree with you that exposing the hash is a little dirty. So maybe we can use tuple or struct to assemble `UUri` + `ListenerWrapper` as key, and it can simplify the design.
Since `UUri` and `ListenerWrapper` are both hashable now, I guess it's possible to use the combination as key. (Need some verification)

Let me know if I misunderstood what you meant.

Reply:
Oh yeah, you may be right. I'll have to try this 🙂

Reply:
Yeah, I think I like this! In the case of `up-client-zenoh-rust` we don't actually need to hang onto the `ListenerWrapper`, but can just use its Hash impl. Pushed a commit with this change.